rfcs/0145-workerpoolid-taskqueueid.md

```diff
@@ -0,0 +1,84 @@
+Old software can be gradually transitioned to use the new concepts internally, accepting both kinds of identifiers externally, and finally drop support for the old identifiers.
+
+## API URLs
+
+In many cases, API URLs have the form `../:provisionerId/:workerType`.
+With some careful configuration of Express with regard to escaping of the `/` character, these API methods can be alterted to interpret the same URLs as containing a `workerPoolId` or `taskQueueId`.
+This means that existing API clients can continue to operate in terms of `provisionerId` and `workerType` while calling the APIs that are defined in terms of `workerPoolId` or `taskQueueId`.
+
+The `https://queue.taskcluster.net/v1/provisioners/:provisionerId/worker-types/:workerType` endpoint is an exception to this rule, as it contains the literal string `/worker-types/` in the middle of the URL.
+However, this endpoint and its friends beginning with `/provisioners` are still experimental and soon will be replaced with worker-manager functionality, at which time they can be deprecated or removed.
+
+EC2-Manager and AWS-Provisioner are also exceptions, but these services are deprecated and will be removed soon.
+
+## Task Definitions
+
+Task definitions currently contain `provisionerId` and `workerType` properties.
+When the transition to this proposal is complete, these properties should be replaced with a single `taskQueueID`.
+
+During the transition, however, the queue must both accomodate input from old and new clients, and produce output that new and old clients can interpret.
+To do so, the `createTask` API method will accept a task definition with either `provisionerId` + `workerType` or `taskQueueId`, translating the former internally into a `taskQueueId`.
+The `task` API method (which returns a task) will return task definitions containing all three fields, allowing interpretation by either old new clients.
+
+## Sunsetting the Old Identifiers
+
+At some point, we would like to relax the requirements on the `workerPoolId` and `taskQueueId` identifiers -- in particular, to stop requiring `/`.
+
+At that point, we can continue to accept tasks in the `createTask` API method containing `provisionerId` and `workerType`.
+These are very likely to continue to exist, especially for small half-forgotten projects with no dedicated CI engineers.
+For such tasks, queue can continue to synthesize a `taskQueueId` containing a `/` character, and so long as workers are still consuming from that queue, nothing will fail.
+
+However, it will not be possible to determine a `provisionerId` and `workerType` for a `taskQueueId` that does not contain a `/`, so queue will stop returning task definitions containing these values.
+This is a breaking API change, but only affects users that *read* task definitions: a smaller population than those who create tasks.
+It is also a fairly simple change to ask those users to adapt to.
```

Comment:
In the worst case, we could continue to return those fields only if the `taskQueueId` matches the previously required regular expression (`^[a-zA-Z0-9-_]{1,38}/[a-z]([-a-z0-9]{0,36}[a-z0-9])?$`).

Reply:
Yeah, we'll have flexibility to see how tricky we want to get.  A downside of the approach you mention is that then some tasks would have the fields while others would not, which might lead to nasty surprises at unexpected times.